### ISSUE TITLE
Issue with translating DataFrame: py_ref_to_r(x) : Unsupported array conversion from 19

### ISSUE THREAD

**Issue by adrivinca** [rstudio/reticulate#248]
Hi reticulate team,

First off, we love your product, and we know this issue will not be terribly easy to diagnose - apologies. We are trying to use reticulate with some pretty complex python bindings of some Java code we have. We've run into an issue getting things back into R that googling hasn't been able to solve. We won't be able to provide a minimal example that is easy to reproduce (unless you feel like installing our package, which has some nontrivial dependencies), but we will do our best to explain it here. 

We have a `Scenario` object with a method `par()` which returns in this case a `pandas.DataFrame` with three columns: 

```
In[11]:  scenario.par('interestrate')

Out[11]: 
  unit  value  year
0    %   0.05  2010
1    %   0.05  2020
2    %   0.05  2030
3    %   0.05  2040

In[12]: type(scenario.par('interestrate'))
Out[12]: pandas.core.frame.DataFrame

In[13]: scenario.par('interestrate').dtypes
Out[13]: 
unit      object
value    float32
year       int32
dtype: object
```

We get the following error using reticulate:

```
> scenario$par("interestrate")
Error in py_ref_to_r(x) : Unsupported array conversion from 19
```

It seems like we're hitting a [case/switch in `python.cpp`](https://github.com/rstudio/reticulate/blob/master/src/python.cpp#L844) where it's trying to translate unicode, but finding no specific case. 

Oddly enough, if we dump this object to csv and read it back in, all works:

foo.py:
```
import pandas as pd

def test_read():
    df =  pd.read_csv('~/test.csv')
    print(df)
    print(df.dtypes)
    return df
```

And this works:
```
import('foo')
foo$test_read()
```

Could you provide any guidance of how we might address this issue? We're happy to do type conversions etc on the Python side.

Cheers

**Comment by adrivinca** [comment]
cc: @gidden

**Comment by gidden** [comment]
And here is the R stacktrace:

```
Error in py_ref_to_r(x): Unsupported array conversion from 19

Traceback:

 

1. scenario$par("interestrate")

2. py_to_r(result)

3. py_to_r.pandas.core.frame.DataFrame(result)

4. py_to_r(x$columns$values)

5. py_to_r.numpy.ndarray(x$columns$values)

6. NextMethod()

7. py_to_r.default(x$columns$values)

8. py_ref_to_r(x)
```

**Comment by jjallaire** [comment]
It looks like we do support unicode objects from NumPy but for some reason this code path isn't picking that up. @kevinushey Could you take a look when you have a chance?

**Comment by gidden** [comment]
Hi Kevin,

I suspect it is an array of scalar arrays of unicode. I'll dig with
@adrivinca and see what we can come up with.

Cheers,
Matt

**Comment by gidden** [comment]
Hi @kevinushey ,

Pickled (and zipped) example here: [ex.zip](https://github.com/rstudio/reticulate/files/1994666/ex.zip)

To confirm, in Python:



```
import pandas as pd
pd.read_pickle('interestrate.pkl')

  | unit | value | year
-- | -- | -- | --
% | 0.05 | 2010
% | 0.05 | 2020
% | 0.05 | 2030
% | 0.05 | 2040
```

With reticulate:

```
library('reticulate')
pd = import('pandas')
pd$read_pickle('interestrate.pkl')

Error in py_ref_to_r(x): Unsupported array conversion from 19
Traceback:

1. pd$read_pickle("interestrate.pkl")
2. py_to_r(result)
3. py_to_r.pandas.core.frame.DataFrame(result)
4. py_to_r(x$columns$values)
5. py_to_r.numpy.ndarray(x$columns$values)
6. NextMethod()
7. py_to_r.default(x$columns$values)
8. py_ref_to_r(x)
```

**Comment by gidden** [comment]
To note, I attempted some very primitive tests, and it appears that something is happening when converting the `year` column (which has dtype `int64`). If I drop that column on the python side, reticulate converts correctly. If I drop others and only try to convert `df[['year']]`, I see the same failure as above.

**Comment by gidden** [comment]
Ok, @kevinushey and @jjallaire, at least I have a viable smoke test here. It looks like the issue is the column name "year". If I rename this (e.g., to "foo") before returning on the Python side, everything converts correctly. I'm not familiar enough with R to know if this is a builtin function that is being called because of name lookup order or what have you.

Since "year" is a standard column name in our data model, I am not sure we could change this easily on our end.. any thoughts on if this can be addressed on the reticulate end?

**Comment by gidden** [comment]
Hi all - final bit of sleuthing here.  `y` below is read in as the pickled DF.

```
In [50]:  y.columns.values
Out [50]:  array(['unit', 'value', 'year'], dtype=object)

In [51]: [type(_) for _ in y.columns.values]
Out [51]: [str, str, numpy.str_]
```

I submitted a minimal PR which fixes this for me, but I have no idea if it affects other use cases.. I wasn't able to find instructions on how to run the test suite, apologies.

**Comment by kevinushey** [comment]
Awesome! Thanks for the reproducible example @gidden -- I can reproduce locally with the provided pickled DataFrame.

One other thing I'm seeing is with autocompletion results within RStudio -- e.g. it seems like attempts to list attributes on this DataFrame are failing, with:

```
> py_list_attributes(pdf)
Error in py_list_attributes_impl(x) : 
  TypeError: expected bytes, numpy.str_ found
```

**Comment by kevinushey** [comment]
Alright, here's a minimal example:

```
library('reticulate')

py_run_string("
import numpy as np
npa = np.array([np.str_('a')], dtype = np.object)
")

py$npa
```

This gives the same error you were seeing:

```
> py$npa
Error in py_ref_to_r(x) : Unsupported array conversion from 19
```

**Comment by kevinushey** [comment]
And similar using plain `reticulate` APIs:

```
np <- import("numpy", convert = FALSE)
arr = np$array(list(np$str_('a')), dtype = np$object)
py_to_r(arr)
```

giving

```
> py_to_r(arr)
Error in py_ref_to_r(x) : Unsupported array conversion from 19
```

**Comment by dfalbel** [comment]
Fixed in #257
